### PR TITLE
fix(agent-runtime): bypass proxy for loopback gateways

### PR DESCRIPTION
**Pull request by AtomsH4**
> ### 🚨 Branch strategy — read before opening this PR
>
> The v2 refactor has merged into `main`, so **`main` is the default branch for active development** (v1 and v2 code currently coexist there — expect large, breaking changes).
>
> - **Active development** (features, refactors, optimizations, fixes for the current codebase) → target **`main`** (the default base).
> - **v1 maintenance** (hotfixes and subsequent v1 releases) → branch from and target **`v1`**, _not_ `main`.
>
> A v1 fix does **not** auto-carry to `main`: if the same bug exists on `main`, open a separate forward-port PR targeting `main`. Before touching subsystems being replaced, read `docs/references/data/` and watch for `@deprecated` markers — they flag code being deleted.

### What this PR does

Before this PR:

Claude Code Agent subprocesses inherited Cherry's proxy environment without Chromium's implicit loopback bypass. Requests to local gateways such as `127.0.0.1:23333` could therefore be sent to a remote proxy, while prewarmed subprocesses could also retain stale proxy settings after a proxy switch.

After this PR:

Claude Code Agent environments merge `localhost`, `127.0.0.1`, `::1`, and `[::1]` into both `no_proxy` and `NO_PROXY` while preserving and deduplicating existing rules. Proxy environment fingerprints trigger subprocess rebuilds on the next fresh turn, cached Cherry-injected shell values are ignored, and external model requests continue using the configured proxy.

N/A

### Why we need it and why it was done in this way

Electron/Chromium applies an implicit loopback bypass, but the Claude Code/Node subprocess network stack only sees process environment variables. Applying the normalization when constructing the Agent environment keeps the behavior aligned without changing Electron's global proxy configuration or the normal-chat network stack.

The following tradeoffs were made:

- The automatic bypass is intentionally limited to four exact loopback forms instead of the full `127.0.0.0/8` range or private networks.
- Active Agent turns are not interrupted; a changed proxy fingerprint rebuilds the subprocess at the next fresh-turn reconciliation.
- Only cached shell values that exactly match Cherry's internal proxy markers are removed, preserving different user-configured login-shell proxy values.

The following alternatives were considered:

- Changing Electron's global proxy bypass, which would broaden the fix beyond Agent subprocesses.
- Disabling proxy inheritance for Agent subprocesses, which would incorrectly bypass the proxy for external model requests.
- Changing shared shell-environment caching or global proxy-controller behavior, which would introduce unnecessary cross-cutting infrastructure changes.

Links to places where the discussion took place: [Chromium proxy bypass rules](https://chromium.googlesource.com/chromium/src/+/main/net/docs/proxy.md#implicit-bypass-rules), [OpenAI Codex network proxy](https://github.com/openai/codex/tree/main/codex-rs/network-proxy), [GitHub Actions proxy guidance](https://docs.github.com/en/actions/how-tos/manage-runners/use-proxy-servers)

### Breaking changes

None.

### Special notes for your reviewer

- Targeted Vitest coverage: 5 files, 244 tests passed.
- `pnpm typecheck:node` passed.
- ESLint and Biome checks passed for all six changed TypeScript files.
- The full repository test suite was not run.

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [x] Branch: This PR targets the correct branch — `main` for active development, `v1` for v1 maintenance fixes
- [x] PR: The PR description is expressive enough and will help future contributors
- [x] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [x] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [x] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [x] Documentation: A [user-guide update](https://docs.cherry-ai.com) was considered and is present (link) or not required. Check this only when the PR introduces or changes a user-facing feature or behavior.
- [x] Self-review: I have reviewed my own code (e.g., via [`/gh-pr-review`](/.claude/skills/gh-pr-review/SKILL.md), `gh pr diff`, or GitHub UI) before requesting review from others

### Release note

```release-note
Fix Agent work mode so localhost and loopback gateways bypass configured remote proxies while external model requests continue using them.
```

### CODE REVIEW

**Review by DeJeune — APPROVED** [review]

File: src/main/ai/runtime/claudeCode/agentProxyEnvironment.ts [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+  return result
+}
+
+export const isAgentProxyEnvironmentKey = (key: string): boolean =>
+  AGENT_PROXY_ENVIRONMENT_KEY_SET.has(key) || CHERRY_PROXY_MARKER_KEY_SET.has(key.toUpperCase())
+
+export const mergeAgentLoopbackProxyBypass = (environment: Environment): Record<string, string | undefined> => {
+  const result = { ...environment }
+  if (!AGENT_PROXY_ENDPOINT_KEYS.some((key) => isNonEmpty(environment[key]))) {
```
Comment (vaayne) [inline src/main/ai/runtime/claudeCode/agentProxyEnvironment.ts]: **[P2] Match proxy environment names case-insensitively on Windows**

This active-proxy check only probes the enumerated exact spellings. Windows environment-variable names are case-insensitive, while this JS record lookup is case-sensitive, so a valid `Http_Proxy=http://corp-proxy:8080` remains effective in the child process but is not detected here. We then skip adding loopback `NO_PROXY`, and the fingerprint below also omits the active proxy, so a warmed session may not rebuild when it changes.

Please canonicalize recognized proxy/bypass keys case-insensitively (with deterministic precedence for duplicate case variants) and use the same normalized representation for merging and fingerprinting. A Windows-oriented test with `Http_Proxy`, `Https_Proxy`, and `No_Proxy` would cover this.

File: src/main/ai/runtime/claudeCode/agentProxyEnvironment.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+  'SOCKS_PROXY',
+  'socks_proxy',
+  'grpc_proxy'
+] as const
+
+const AGENT_PROXY_BYPASS_KEYS = ['NO_PROXY', 'no_proxy'] as const
+const AGENT_PROXY_ENVIRONMENT_KEYS = [...AGENT_PROXY_ENDPOINT_KEYS, ...AGENT_PROXY_BYPASS_KEYS] as const
+const AGENT_PROXY_ENVIRONMENT_KEY_SET = new Set<string>(AGENT_PROXY_ENVIRONMENT_KEYS)
+const CHERRY_PROXY_MARKER_KEY_SET = new Set([CHERRY_NODE_PROXY_RULES_ENV, CHERRY_NODE_PROXY_BYPASS_RULES_ENV])
+const LOOPBACK_BYPASS_RULES = ['localhost', '127.0.0.1', '::1', '[::1]'] as const
```
Comment (vaayne) [inline src/main/ai/runtime/claudeCode/agentProxyEnvironment.ts:32]: **[P2] Add the materialized gateway host instead of a fixed loopback spelling list**

The API gateway host is configurable, but this list only bypasses four textual forms. For example, with `feature.api_gateway.host=127.0.0.2` and an active `HTTP_PROXY`, the generated `NO_PROXY` does not match the local gateway, so its HTTP requests (including gateway credentials and agent traffic) can still be sent to the proxy. Other valid loopback forms such as `127.1`, expanded IPv6 loopback, and mapped loopback are also missed.

Please derive/canonicalize the host from the materialized gateway URL and add that actual destination to both bypass variables; ideally also recognize the full loopback address ranges. Add focused cases for at least `127.0.0.2` and alternate IPv6 loopback forms.

File: src/main/ai/runtime/claudeCode/agentProxyEnvironment.ts [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+  const result = { ...environment }
+  const cherryProxyUrl = result[CHERRY_NODE_PROXY_RULES_ENV]
+  const cherryBypassRules = result[CHERRY_NODE_PROXY_BYPASS_RULES_ENV]
+
+  delete result[CHERRY_NODE_PROXY_RULES_ENV]
+  delete result[CHERRY_NODE_PROXY_BYPASS_RULES_ENV]
+
+  if (cherryProxyUrl !== undefined) {
+    for (const key of AGENT_PROXY_ENDPOINT_KEYS) {
+      if (result[key] === cherryProxyUrl) {
```
Comment (vaayne) [inline src/main/ai/runtime/claudeCode/agentProxyEnvironment.ts]: **[P2] Value equality cannot distinguish a stale Cherry proxy from a user-owned shell proxy**

This deletes every proxy variable whose value equals the Cherry marker. A user may explicitly export the same common local-proxy URL in their shell profile (for example `HTTP_PROXY=http://127.0.0.1:7890`) while Cherry is configured to use that URL too. After Cherry's proxy is disabled, the cached login-shell value is removed even though the user's shell still intentionally owns it, and the agent unexpectedly loses its proxy.

Please avoid inferring provenance from value equality. One robust approach is to capture the login shell with Cherry-managed proxy variables/markers removed from the shell process's inherited environment; values explicitly set by startup files then reappear and can be preserved. Add a regression test where the shell explicitly produces the same URL as the Cherry marker.

**Review by vaayne — CHANGES_REQUESTED** [review]
The loopback-proxy fix still has three edge cases that can leave local gateway traffic proxied or remove a user-owned proxy. Focused review tests passed (3 files / 166 tests), but these scenarios are not covered. Details are inline.

File: src/main/ai/runtime/claudeCode/agentProxyEnvironment.ts (hunk 1 above)
Comment (AtomsH4) [reply src/main/ai/runtime/claudeCode/agentProxyEnvironment.ts]: Fixed in f53978fa28. On Windows, recognized proxy and bypass keys are now canonicalized case-insensitively using deterministic precedence that mirrors Node child_process: raw environment keys are lexicographically sorted and the first case-insensitive variant wins. Bypass merging, proxy-key classification, and the rebuild fingerprint all use the same effective representation. Added coverage for mixed-case Http_Proxy and No_Proxy, plus duplicate-case endpoint and bypass precedence.

File: src/main/ai/runtime/claudeCode/agentProxyEnvironment.ts (hunk 2 above)
Comment (AtomsH4) [reply src/main/ai/runtime/claudeCode/agentProxyEnvironment.ts:32]: Fixed in f53978fa28. The Agent runtime now extracts the hostname from the materialized gateway URL and appends that exact destination to both NO_PROXY and no_proxy; the same route-aware value is included in rebuild fingerprint derivation so warm and current signatures stay aligned. Added a focused 127.0.0.2 unit case and a materialized gateway integration case. I intentionally used the exact selected gateway host rather than bypassing the full 127/8 range, which keeps the exception scoped to Cherry's actual local destination and avoids silently bypassing user proxy policy for unrelated loopback services.

File: src/main/ai/runtime/claudeCode/agentProxyEnvironment.ts (hunk 3 above)
Comment (AtomsH4) [reply src/main/ai/runtime/claudeCode/agentProxyEnvironment.ts]: Fixed in f53978fa28. The login-shell cleanup now removes only Cherry's internal marker variables and never deletes proxy or bypass variables based on equal values. That preserves a user-owned HTTP_PROXY or NO_PROXY even when it matches Cherry's former setting. Added a regression test for this equality case and updated rebuild and settings tests to make the provenance ambiguity explicit. The tradeoff is that an ambiguous stale cached value is preserved rather than risking deletion of user configuration.

File: src/main/ai/runtime/claudeCode/agentProxyEnvironment.ts [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+  }
+
+  return rules
+}
+
+/** Remove internal metadata only; equal values cannot prove that Cherry owns a shell proxy variable. */
+export const stripInheritedCherryProxyMarkers = (environment: Environment): Record<string, string | undefined> => {
+  const result = { ...environment }
+
+  delete result[CHERRY_NODE_PROXY_RULES_ENV]
```
Comment (vaayne) [inline src/main/ai/runtime/claudeCode/agentProxyEnvironment.ts:90]: **[P2] Do not retain a cached Cherry proxy after Cherry proxying is disabled**

Removing only the markers fixes the false-positive deletion of a user-owned proxy, but it also preserves proxy endpoints that were injected by Cherry and captured by the memoized `getShellEnv()`. A concrete sequence is: enable Cherry proxy -> the login shell inherits `HTTP_PROXY` plus these markers and that environment is cached -> disable Cherry proxy (the controller clears current `process.env`) -> build a later Agent environment. This function drops the only provenance metadata but retains the cached endpoint, so the Agent continues using the disabled/stale proxy; the rebuild fingerprint also continues to include it. The updated tests currently encode that stale endpoint retention.

Please preserve provenance without relying on value equality—for example, prevent Cherry-managed variables from entering the durable login-shell capture, cache a clean raw shell environment and merge only the current Cherry proxy at spawn time, or invalidate/re-capture the shell cache when proxy configuration changes. The settings builder and fingerprint path should consume the same provenance-aware snapshot.

**Review by vaayne — CHANGES_REQUESTED** [review]
The latest hardening fixes the three previously reported cases, but the provenance change leaves one opposite-direction regression: a Cherry proxy captured in the memoized login-shell environment survives after that proxy is disabled. Details inline.

File: src/main/ai/runtime/claudeCode/agentProxyEnvironment.ts (hunk 4 above)
Comment (AtomsH4) [reply src/main/ai/runtime/claudeCode/agentProxyEnvironment.ts:90]: Fixed in 14b9eec334. When the memoized login-shell snapshot contains Cherry proxy markers that no longer match the current proxy configuration, the Agent runtime now refreshes that snapshot before stripping the markers. Both settings construction and rebuild fingerprinting consume the same provenance-aware snapshot and the same current proxy snapshot. This removes disabled Cherry-managed proxy values while preserving user-owned values re-created by shell startup files, including values equal to the former Cherry proxy URL. Matching markers retain the memoized snapshot, so unchanged configurations do not add another shell spawn.

Validation: 3 focused files / 180 tests passed; lint, Node/Web/AiCore typechecks, i18n, format, and documentation-link checks passed. I also manually reproduced failed proxy enabled -> switch Proxy Mode to No Proxy -> send an Agent turn: ShellEnv was recaptured, the gateway returned 200 for /v1/messages, the mock model received the streaming request, and the Agent response completed without a new request reaching the stale proxy. The full test suite was intentionally not run.

**Review by vaayne — APPROVED** [review]

**Review by vaayne — APPROVED** [review]
LGTM
